app/controllers/api/v1/service_plans_controller.rb

```diff
@@ -6,21 +8,59 @@ class ServicePlansController < ApplicationController
 
       def order
         service_plan = model.find(params_for_order[:service_plan_id].to_i)
-        task = Task.create!(:tenant => service_plan.tenant, :state => "pending", :status => "ok")
 
-        messaging_client.publish_message(
-          :service => "platform.topological-inventory.operations-openshift",
-          :message => "ServicePlan.order",
-          :payload => payload_for_order(task, service_plan)
-        )
+        retrieve_source_type(service_plan) do |source_type|
+          task = Task.create!(:tenant => service_plan.tenant, :state => "pending", :status => "ok")
+
+          messaging_client.publish_message(
+            :service => "platform.topological-inventory.operations-#{source_type.name}",
+            :message => "ServicePlan.order",
+            :payload => payload_for_order(task, service_plan)
+          )
+
+          render :json => {:task_id => task.id}
+        end
 
-        render :json => {:task_id => task.id}
       rescue ActiveRecord::RecordNotFound
         head :bad_request
+      rescue StandardError => err
+        error_document = ManageIQ::API::Common::ErrorDocument.new.add(err.message)
+        render :json => error_document.to_h, :status => error_document.status
       end
 
       private
 
+      def sources_api_client
+        @sources_api_client ||=
```

Comment:
I don't think you can cache this without including the x-rh-identity context, otherwise the tenant from the first API call will be used for every subsequent call.